src/casp/prepare_for_training.py

```diff
@@ -0,0 +1,135 @@
+"""
+Prepare data for extract by randomizing, preprocessing and staging in temporary tables
+"""
+import argparse
+import time
+
+from google.cloud import bigquery
+
+
+def execute_query(client, sql):
+    """
+    Runs the supplied query
+    """
+    start = time.time()
+
+    # print(f"RUNNING:\n {sql}")
+    query = client.query(sql)
+    results = query.result()
+
+    job = client.get_job(query.job_id)
+    gb_billed = int(0 if job.total_bytes_billed is None else job.total_bytes_billed) / (1024 * 1024 * 1024)
+    print(f"COMPLETED ({time.time() - start} seconds, {gb_billed} GBs scanned)")
+
+    return results
+
+
+def prepare_feature_summary(client, project, dataset, extract_table_prefix):
+    """
+    create feature/gene level summary -- scan of full dataset
+    """
+    sql = f"""
+        CREATE OR REPLACE TABLE `{project}.{dataset}.{extract_table_prefix}__extract_feature_summary`
+        AS
+        SELECT  f.feature_name,
+                SUM(m.raw_counts) total_raw_counts,
+                COUNT(distinct CASE WHEN m.raw_counts > 0 THEN m.cas_cell_index ELSE null END) cells_with_counts
+        FROM `{project}.{dataset}.cas_raw_count_matrix` m
+        JOIN `{project}.{dataset}.cas_feature_info` f ON (m.cas_feature_index = f.cas_feature_index)
+        GROUP BY f.feature_name
+    """
+    print("Creating Feature Summary...")
+    query = execute_query(client, sql)
+    return query
+
+
+def prepare_feature_info(client, project, dataset, extract_table_prefix, min_observed_cells):
+    """
+    create subset of features for extract with new local identifiers
```

Comment:
I'll update the docstring, but here's what should be happening (so please check the logic!)

the creation of `{extract_table_prefix}__extract_feature_summary` has already grouped by `feature_name` so that table has a unique set of `feature_name`

Here we are subsetting that, and creating a new (local) index number which is just the index of the sorted features by name.  The comment should def say "new index" rather than "local identifiers"